code/game/machinery/hologram.dm

```diff
@@ -90,175 +245,175 @@ var/list/holopads = list()
-/obj/machinery/hologram/holopad/hear_talk(mob/living/M, text, verb, datum/language/speaking)
-	if(M && hologram && master)//Master is mostly a safety in case lag hits or something.
-		master.relay_speech(M, text, verb, speaking)
-
-/obj/machinery/hologram/holopad/hear_message(mob/living/M, text)
-	if(M&&hologram&&master)//Master is mostly a safety in case lag hits or something.
-		var/name_used = M.GetVoice()
-		var/rendered = "<i><span class='game say'>Holopad received, <span class='name'>[name_used]</span> [text]</span></i>"
-		master.show_message(rendered, 2)
-	return
-
-/obj/machinery/hologram/holopad/proc/create_holo(mob/living/silicon/ai/A, turf/T = loc)
-	hologram = new(T)//Spawn a blank effect at the location.
-	hologram.icon = A.holo_icon
-	hologram.mouse_opacity = 0//So you can't click on it.
-	hologram.layer = FLY_LAYER//Above all the other objects/mobs. Or the vast majority of them.
-	hologram.anchored = 1//So space wind cannot drag it.
-	hologram.name = "[A.name] (Hologram)"//If someone decides to right click.
-	hologram.set_light(2)	//hologram lighting
-	set_light(2)			//pad lighting
-	icon_state = "holopad1"
-	A.holo = src
-	master = A//AI is the master.
-	use_power = 2//Active power usage.
-	return 1
+/obj/machinery/hologram/holopad/process()
+	for(var/I in masters)
+		var/mob/living/master = I
+		var/mob/living/silicon/ai/AI = master
+		if(!istype(AI))
+			AI = null
+
+		if(!qdeleted(master) && !master.incapacitated() && master.client && (!AI || AI.eyeobj))//If there is an AI attached, it's not incapacitated, it has a client, and the client eye is centered on the projector.
+			if(!(stat & NOPOWER))//If the  machine has power
+				if(AI)	//ais are range based
+					if(HOLOPAD_MODE == RANGE_BASED)
+						if(get_dist(AI.eyeobj, src) <= holo_range)
+							continue
+
+					if(!(AI.current))//if the ai jumped to core
+						clear_holo(AI)
+						return 1
+
+					var/obj/machinery/hologram/holopad/pad_close = get_closest_atom(/obj/machinery/hologram/holopad, holopads, AI.eyeobj)
+					if(get_dist(pad_close, AI.eyeobj) <= pad_close.holo_range)
+						clear_holo(AI)
+						pad_close.activate_holo(AI, 1)
+				else
+					continue
+
+		clear_holo(master)//If is a non AI holo clear it.
+
+	if(outgoing_call)
+		outgoing_call.Check()
+
+	for(var/I in holo_calls)
+		var/datum/holocall/HC = I
+		//Sanity check and skip if no longer valid call
+		if(!HC.Check())
+			atom_say("Call was terminated at remote terminal.")
+			continue
+
+		if(HC.connected_holopad != src)
+			if(force_answer_call && world.time > (HC.call_start_time + (HOLOPAD_MAX_DIAL_TIME / 2)))
+				HC.Answer(src)
+				break
+			if(outgoing_call)
+				HC.Disconnect(src)//can't answer calls while calling
+			else
+				playsound(src, 'sound/machines/twobeep.ogg', 100)	//bring, bring!
+
+/obj/machinery/hologram/holopad/proc/move_hologram(mob/living/user, turf/new_turf)
+	if(masters[user])
+		var/obj/effect/overlay/holo_pad_hologram/H = masters[user]
+		step_to(H, new_turf)

```

Comment:
well by my testing just now @tigercat2000 its making the sprite face the right direction if a holopad changes (IE the AI) when the hologram moves around the area (possible for players too) i am welcome to another way to do it..i also think i put it in to prevent an odd snap back issue when the AI moved to another holopad the hologram icon would center on the holopad and not the ai eye location until the eye moved.